lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,198 @@
+require 'uri'
+
+module ProxyAPI
+  class ExternalIpam < ProxyAPI::Resource
+    def initialize(args)
+      @url = args[:url] + "/ipam"
+      super args
+    end
+
+    # Queries External IPAM and retrieves the next available IP address for the given subnet.
+    # The IP returned is NOT reserved in External IPAM database. It is however written to an in-memory,
+    # thread safe cache, on the proxy side, with the subnet CIDR/mac address as the key(see
+    # IP cache structure below). This will prevent the same IP being suggested for different mac addresses,
+    # and will handle race condition scenarios where multiple hosts are being provisioned simultaneously.
+    #
+    # The IP address is only actually reserved in the External IPAM database upon successful host
+    # and/or interface creation.
+    #
+    # In-memory IP cache structure:
+    # ===============================
+    #   "100.55.55.0/24":{
+    #      "00:0a:95:9d:68:10": {"ip": "100.55.55.1", "timestamp": "2019-09-17 12:03:43 -D400"}
+    #   },
+    #   "123.11.33.0/24":{
+    #      "00:0a:95:9d:68:33": {"ip": "123.11.33.1", "timestamp": "2019-09-17 12:04:43 -0400"}
+    #      "00:0a:95:9d:68:34": {"ip": "123.11.33.2", "timestamp": "2019-09-17 12:05:48 -0400"}
+    #      "00:0a:95:9d:68:35": {"ip": "123.11.33.3", "timestamp:: "2019-09-17 12:06:50 -0400"}
+    #   }
+    # }
+    #
+    # Params: 1. subnet:  The Foreman subnet object
+    #         2. mac:     The mac address of the interface obtaining the IP address for
+    #
+    # Returns: A hash with next available IP in the "data" field, or a hash with "error" on failure
+    #
+    # Examples:
+    #   Response if success:
+    #     {"code": 200, "success": true, "data": "100.55.55.3", "time": 0.012}
+    #   Response if missing required params:
+    #     {"error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","A 'mac' address must be provided(e.g. 00:0a:95:9d:68:10)"]}
+    #   Response if subnet error:
+    #     {"error": "The specified subnet does not exist in External IPAM."}
+    #   Response if there are no free addresses:
+    #     {"error": "No free addresses found"}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def next_ip(subnet, mac)
+      raise "subnet.network cannot be nil" unless subnet.network.present?
+      parse get("/next_ip?cidr=#{subnet.network_address}&mac=#{mac}")
```

Comment:
Will this API also work for IPv6? Can you add few words in the documentation comment above? If it does not work, you need to handle this in the method and call the correct endpoint in case the input is IPv6.

Also define some error in the foreman-smartproxy protocol which is raised when IPv6 is not supported (or vice versa). We will likely see IPAM implementations which are IPv4 only (some legacy systems).